modules/frontend/querysharding.go

```diff
@@ -38,36 +41,156 @@ func ShardingWare(queryShards int, logger log.Logger) Middleware {
 	logger          log.Logger
 	blockBoundaries [][]byte
 }
 
-// Do implements Handler
-func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+// RoundTrip implements http.RoundTripper
+func (s shardQuery) RoundTrip(r *http.Request) (*http.Response, error) {
 	ctx := r.Context()
 	span, ctx := opentracing.StartSpanFromContext(ctx, "frontend.ShardQuery")
 	defer span.Finish()
 
 	// context propagation
 	r = r.WithContext(ctx)
+	reqs, err := s.buildShardedRequests(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute requests
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var overallTrace *tempopb.Trace
+	var overallError error
+	statusCode := http.StatusNotFound
+	statusMsg := "trace not found"
+
+	for _, req := range reqs {
+		wg.Add(1)
+		go func(innerR *http.Request) {
+			defer wg.Done()
+
+			resp, err := s.next.RoundTrip(innerR)
+
+			mtx.Lock()
+			defer mtx.Unlock()
+			if err != nil {
+				overallError = err
+			}
+
+			if shouldQuit(r.Context(), statusCode, overallError) {
+				return
+			}
+
+			// check http error
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "error querying proxy target", "url", innerR.RequestURI, "err", err)
+				overallError = err
+				return
+			}
+
+			// if not found bail
+			if resp.StatusCode == http.StatusNotFound {
+				return
+			}
+
+			// if the status code is anything but happy, save the error and pass it down the line
+			if resp.StatusCode != http.StatusOK {
+				// todo: if we cancel the parent context here will it shortcircuit the other queries and fail fast?
+				statusCode = resp.StatusCode
+				bytesMsg, err := io.ReadAll(resp.Body)
+				if err != nil {
+					_ = level.Error(s.logger).Log("msg", "error reading response body status != ok", "url", innerR.RequestURI, "err", err)
+				}
+				statusMsg = string(bytesMsg)
+				return
+			}
+
+			// successful query, read the body
+			buff, err := io.ReadAll(resp.Body)
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "error reading response body status == ok", "url", innerR.RequestURI, "err", err)
+				overallError = err
+				return
+			}
+
+			// marshal into a trace to combine.
+			// todo: better define responsibilities between middleware. the parent middleware in frontend.go actually sets the header
+			//  which forces the body here to be a proto encoded tempopb.Trace{}
+			trace := &tempopb.Trace{}
+			err = proto.Unmarshal(buff, trace)
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "error unmarshalling response", "url", innerR.RequestURI, "err", err, "body", string(buff))
+				overallError = err
+				return
+			}
+
+			// happy path
+			statusCode = http.StatusOK
+			overallTrace, _, _, _ = model.CombineTraceProtos(overallTrace, trace)
+		}(req)
+	}
+	wg.Wait()
+
+	if overallError != nil {
+		return nil, overallError
+	}
+
+	if overallTrace == nil || statusCode != http.StatusOK {
+		// translate non-404s into 500s. if, for intance, we get a 400 back from an internal component
```

Comment:
Small typo
```suggestion
		// translate non-404s into 500s. if, for instance, we get a 400 back from an internal component
```

Also, I appreciate the comment explaining why we return 500 :+1: